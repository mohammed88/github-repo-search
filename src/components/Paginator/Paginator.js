import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Link from '../../ui/Link/';
import {darkGray} from '../../ui/styles/colors';
import {Text14} from '../../ui/styles/typography';

const Wrapper = styled.div`
	align-items: center;
	display: flex;
	justify-content: space-between;
	width: 300px;
	${Text14};
`;

export const PaginatorLink = styled(Link)`
	color: ${props => props.disabled ? darkGray : "default"};
	flex: 0 1 0;
	pointer-events: ${props => props.disabled ? "none" : "auto"};
	${Text14};
`;

class Paginator extends React.Component {
	state = {
		currentPage: 1,
		prevPropsCurrentPage: null,
	};

	static getDerivedStateFromProps(props, state) {
		if(
			state.prevPropsCurrentPage !== props.currentPage && 
			props.currentPage <= props.totalCount
		) {
			return {
				currentPage: props.currentPage,
				prevPropsCurrentPage: props.currentPage,
			};
		}

		return state;
	}

	render() {
		const {currentPage} = this.state;
		const {totalCount, className} = this.props;

		return (
			<Wrapper className={className}>
				<PaginatorLink 
					onClick={this.handleFirst} 
					disabled={currentPage === 1}
				>
					First
				</PaginatorLink>
				<PaginatorLink
					onClick={this.handlePrev} 
					disabled={currentPage === 1}
				>
					Previous
				</PaginatorLink>

				<div>{currentPage} of {totalCount}</div>

				<PaginatorLink 
					onClick={this.handleNext}
					disabled={currentPage === totalCount}
				>
					Next
				</PaginatorLink>
				<PaginatorLink 
					onClick={this.handleLast}
					disabled={currentPage === totalCount}
				>
					Last
				</PaginatorLink>
			</Wrapper>
		);
	}

	handleNext = (e) => {
		const currentPage = Math.min(this.state.currentPage + 1, this.props.totalCount);
		this.setState({currentPage}, this.props.onChange.bind(null, currentPage));
	};

	handlePrev = () => {
		const currentPage = Math.max(this.state.currentPage - 1, 1);
		this.setState({currentPage}, this.props.onChange.bind(null, currentPage));
	};

	handleFirst = () => this.setState(
		{currentPage: 1}, 
		this.props.onChange.bind(null, 1)
	);

	handleLast = () => this.setState({
		currentPage: this.props.totalCount}, 
		this.props.onChange.bind(null, this.props.totalCount)
	);
}

Paginator.propTypes = {
	totalCount: PropTypes.number.isRequired,
};

Paginator.defaultProps = {
	currentPage: 1,
	onChange: () => {},
};

export default Paginator;