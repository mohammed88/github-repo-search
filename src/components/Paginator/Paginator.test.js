import React from 'react';
import {shallow} from 'enzyme';

import Paginator, {PaginatorLink} from './Paginator';

describe('Paginator', () => {
	it('should render', () => {
		const wrapper = shallow(<Paginator totalCount={10} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should show current page passed as prop', () => {
		const currPage = 5;
		const wrapper = shallow(<Paginator totalCount={10} currentPage={currPage} />)

		expect(wrapper.state().currentPage).toEqual(currPage);
	});

	it('should show current page as 1 if currentPage > totalCount', () => {
		const currPage = 30;
		const wrapper = shallow(<Paginator totalCount={10} currentPage={currPage} />)

		expect(wrapper.state().currentPage).toEqual(1);
	});

	it('should increment page count', () => {
		const wrapper = shallow(<Paginator totalCount={10} />);

		wrapper.find(PaginatorLink).at(2).simulate('click');

		expect(wrapper.state().currentPage).toEqual(2);
	});

	it('should decrement page count', () => {
		const currPage = 5;
		const wrapper = shallow(<Paginator totalCount={10} currentPage={currPage} />);

		wrapper.find(PaginatorLink).at(1).simulate('click');

		expect(wrapper.state().currentPage).toEqual(currPage - 1);
	});

	it('should set page count to 1', () => {
		const currPage = 5;
		const wrapper = shallow(<Paginator totalCount={10} currentPage={currPage} />);

		wrapper.find(PaginatorLink).at(0).simulate('click');

		expect(wrapper.state().currentPage).toEqual(1);
	});

	it('should set page count to totalCount', () => {
		const currPage = 5;
		const totalCount = 10;
		const wrapper = shallow(<Paginator totalCount={totalCount} currentPage={currPage} />);

		wrapper.find(PaginatorLink).at(3).simulate('click');

		expect(wrapper.state().currentPage).toEqual(totalCount);
	});

	it('should trigger onChange handler', () => {
		const currPage = 5;
		const totalCount = 10;
		const mockCallback = jest.fn();
		const wrapper = shallow(<Paginator totalCount={totalCount} onChange={mockCallback} />);

		wrapper.find(PaginatorLink).at(3).simulate('click');
		
		expect(mockCallback.mock.calls.length).toBe(1);
	});
});