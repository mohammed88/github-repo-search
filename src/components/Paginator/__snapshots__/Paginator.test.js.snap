// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Paginator
    currentPage={1}
    onChange={[Function]}
    totalCount={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          disabled={true}
          onClick={[Function]}
        >
          First
        </ForwardRef>,
        <ForwardRef
          disabled={true}
          onClick={[Function]}
        >
          Previous
        </ForwardRef>,
        <div>
          1
           of 
          10
        </div>,
        <ForwardRef
          disabled={false}
          onClick={[Function]}
        >
          Next
        </ForwardRef>,
        <ForwardRef
          disabled={false}
          onClick={[Function]}
        >
          Last
        </ForwardRef>,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "First",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "First",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": false,
            "rules": Array [
              "
	",
              "
  color: ",
              "#EA7F28",
              ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
              "
",
              "
	color: ",
              [Function],
              ";
	flex: 0 1 0;
	pointer-events: ",
              [Function],
              ";
	",
              "
	font-family: ",
              "'Roboto', sans-serif",
              ";
	font-size: 14px;
",
              ";
",
            ],
          },
          "displayName": "Styled(styled.a)",
          "foldedComponentIds": Array [
            "sc-bdVaJa",
          ],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Previous",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Previous",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": false,
            "rules": Array [
              "
	",
              "
  color: ",
              "#EA7F28",
              ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
              "
",
              "
	color: ",
              [Function],
              ";
	flex: 0 1 0;
	pointer-events: ",
              [Function],
              ";
	",
              "
	font-family: ",
              "'Roboto', sans-serif",
              ";
	font-size: 14px;
",
              ";
",
            ],
          },
          "displayName": "Styled(styled.a)",
          "foldedComponentIds": Array [
            "sc-bdVaJa",
          ],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            1,
            " of ",
            10,
          ],
        },
        "ref": null,
        "rendered": Array [
          1,
          " of ",
          10,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Next",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Next",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": false,
            "rules": Array [
              "
	",
              "
  color: ",
              "#EA7F28",
              ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
              "
",
              "
	color: ",
              [Function],
              ";
	flex: 0 1 0;
	pointer-events: ",
              [Function],
              ";
	",
              "
	font-family: ",
              "'Roboto', sans-serif",
              ";
	font-size: 14px;
",
              ";
",
            ],
          },
          "displayName": "Styled(styled.a)",
          "foldedComponentIds": Array [
            "sc-bdVaJa",
          ],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Last",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Last",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": false,
            "rules": Array [
              "
	",
              "
  color: ",
              "#EA7F28",
              ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
              "
",
              "
	color: ",
              [Function],
              ";
	flex: 0 1 0;
	pointer-events: ",
              [Function],
              ";
	",
              "
	font-family: ",
              "'Roboto', sans-serif",
              ";
	font-size: 14px;
",
              ";
",
            ],
          },
          "displayName": "Styled(styled.a)",
          "foldedComponentIds": Array [
            "sc-bdVaJa",
          ],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "a",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bxivhb",
        "isStatic": true,
        "rules": Array [
          "
	align-items: center;
	display: flex;
	justify-content: space-between;
	width: 300px;
	",
          "
	font-family: ",
          "'Roboto', sans-serif",
          ";
	font-size: 14px;
",
          ";
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bxivhb",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            disabled={true}
            onClick={[Function]}
          >
            First
          </ForwardRef>,
          <ForwardRef
            disabled={true}
            onClick={[Function]}
          >
            Previous
          </ForwardRef>,
          <div>
            1
             of 
            10
          </div>,
          <ForwardRef
            disabled={false}
            onClick={[Function]}
          >
            Next
          </ForwardRef>,
          <ForwardRef
            disabled={false}
            onClick={[Function]}
          >
            Last
          </ForwardRef>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "First",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "First",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": false,
              "rules": Array [
                "
	",
                "
  color: ",
                "#EA7F28",
                ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
                "
",
                "
	color: ",
                [Function],
                ";
	flex: 0 1 0;
	pointer-events: ",
                [Function],
                ";
	",
                "
	font-family: ",
                "'Roboto', sans-serif",
                ";
	font-size: 14px;
",
                ";
",
              ],
            },
            "displayName": "Styled(styled.a)",
            "foldedComponentIds": Array [
              "sc-bdVaJa",
            ],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Previous",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Previous",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": false,
              "rules": Array [
                "
	",
                "
  color: ",
                "#EA7F28",
                ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
                "
",
                "
	color: ",
                [Function],
                ";
	flex: 0 1 0;
	pointer-events: ",
                [Function],
                ";
	",
                "
	font-family: ",
                "'Roboto', sans-serif",
                ";
	font-size: 14px;
",
                ";
",
              ],
            },
            "displayName": "Styled(styled.a)",
            "foldedComponentIds": Array [
              "sc-bdVaJa",
            ],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              1,
              " of ",
              10,
            ],
          },
          "ref": null,
          "rendered": Array [
            1,
            " of ",
            10,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Next",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Next",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": false,
              "rules": Array [
                "
	",
                "
  color: ",
                "#EA7F28",
                ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
                "
",
                "
	color: ",
                [Function],
                ";
	flex: 0 1 0;
	pointer-events: ",
                [Function],
                ";
	",
                "
	font-family: ",
                "'Roboto', sans-serif",
                ";
	font-size: 14px;
",
                ";
",
              ],
            },
            "displayName": "Styled(styled.a)",
            "foldedComponentIds": Array [
              "sc-bdVaJa",
            ],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Last",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Last",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": false,
              "rules": Array [
                "
	",
                "
  color: ",
                "#EA7F28",
                ";
  font-size: 1rem;
  text-decoration: none;

  :hover {
  	text-decoration: underline;
    cursor: pointer;
  }
",
                "
",
                "
	color: ",
                [Function],
                ";
	flex: 0 1 0;
	pointer-events: ",
                [Function],
                ";
	",
                "
	font-family: ",
                "'Roboto', sans-serif",
                ";
	font-size: 14px;
",
                ";
",
              ],
            },
            "displayName": "Styled(styled.a)",
            "foldedComponentIds": Array [
              "sc-bdVaJa",
            ],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "a",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bxivhb",
          "isStatic": true,
          "rules": Array [
            "
	align-items: center;
	display: flex;
	justify-content: space-between;
	width: 300px;
	",
            "
	font-family: ",
            "'Roboto', sans-serif",
            ";
	font-size: 14px;
",
            ";
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bxivhb",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
